contributing.rst

```diff
@@ -77,3 +77,63 @@ Developer Guidelines
     in C4.1 (e.g., imperative phrasing, wrap lines at 72 characters).
 
 For more details about C4.1, including commit requirements, see `RFC 1 <https://github.com/flux-framework/rfc/blob/master/spec_1.rst>`_.
+
+-------
+Testing
+-------
+
+Testing is an important part of development throughout all of the Flux
+components, and any changes or additions contributed should come with adequate
+testing. The below examples show how to run tests in a Flux project within
+their respective ``t/`` directories.
+
+To run a single test within the test harness, specify the test file on the
+command line:
+
+.. code-block:: console
+
+  $ ./t0000-testname.t -d -v
+
+Python tests can also be run directly, but to get the correct ``PYTHONPATH``
+for your build directory, run under ``flux(1)``:
+
+.. code-block:: console
+
+  $ ../src/cmd/flux ./python/t0000-test.py
+
+or:
+
+.. code-block:: console
+
+  $ src/cmd/flux start
+  $ ./python/t000-testname.py
+
+Other Helpful Testsuite Tips
+----------------------------
+
+* Use the ``make check-prep`` target at the top-level build directory of
+  flux-core to make everything necessary for running tests, but without running
+  any of the tests under ``./t``. This is useful if you then want to run only a
+  specific test, or run all tests as Flux jobs, e.g.
+  ``flux mini bulksubmit --watch --progress ./{} ::: t[0-9]*.t python/t*.py lua/*.t``
+
+* Running all tests in Flux takes some time. If you are developing a feature
+  and only a few tests are failing make recheck can be useful, it will only
+  re-run failing tests. This makes for a more efficient develop-test-debug 
+  cycle.
+
+*  Many sharness tests re-run themselves under a Flux instance using the
+   ``test_under_flux`` command. For these tests you can set
+   ``FLUX_TEST_VALGRIND=t`` and all flux-brokers in the test instance will be
+   run under valgrind for memory debugging. (This is also useful to tease out
+   race conditions in your test code), e.g.:
```

Comment:
`test_under_flux` is actually a shell function not a command.

Also, I think `test_under_flux` is not a prereq for `FLUX_TEST_VALGRIND`.

Reply:
Oh cool!! Thanks for pointing that out.

Perhaps I should separate that note about running tests under `valgrind` as its own bullet point.